_posts/2019-02-25-practice-oop-to-front-end-universal-state-module.md

```diff
@@ -0,0 +1,173 @@
+
+## Motivation
+
+
+As front-end single-page application development becomes increasingly complex, we have to use some state management or state containers (collectively referred to as state library) in order to develop complex apps, and we need a model design that is easier to modularize.
+
+State management libraries are quite abundant in the front-end space, there are [Redux](https://github.com/reduxjs/redux), [MobX](https://github.com/mobxjs/mobx), [Vuex](https://github.com/vuejs/vuex), and the self-contained state management that comes with [Angular](https://github.com/angular/angular) to name a few. Redux is a predictable state container with immutable data structure. MobX is a observable state management. Vuex is centralized state management with observable pattern for Vue.js. As for modularization, Angular has its own implementations already, but the rest of the state management libraries only started to deal with this new requirement in complex systems in recent years.
+
+In this article, let's explore an OOP modular design that has universal support for popular state management libraries.
+
+## Universal state module
+
+Commonly, object-oriented programming (OOP) is used in the architecture design of large projects in the front-end. The following questions are often asked when deciding on a state management library:
+
+  * Is it Redux or MobX more suitable for React?
+  * Is Redux suitable for OOP?
+  * What are the pros and cons of using MobX's observable in React?
+  * How to design the OOP with Vuex?
+
+In addition, Universal JavaScript is more for JavaScript's perform environment in most cases, and once a architecture design selects a state library, it will mean that this architecture is difficult to disengage from the use of this state library, and any system based on this architecture will be based on such a state library. But a better front-end architecture would include more flexible options and scalability, especially the front-end parts of generic requirements such as typical integration business, which can be reflected in the choice of view render libraries and even the availability of state libraries, such as in mainstream scenarios React+Redux/React+MobX/Vue+Vuex/Angular and so on have a choice, then it brings the problem is how to solve universal state module.
+
+## Specific issues that need to be solved
+
+
+* The design of OOP based on state libraries such as Redux/MobX/Vuex, which is also the most important, especially for React/Vue.
+* Whether the encapsulated OOP design is simple enough and easy to use, while they are quite flexible.
+* From DDD's way, the dependencies between complex domain modules require IoC, and the startup logic between them is dependent, so there must be a similar event mechanism or the introduction of the module lifecycle.
+
+> to solve these issues, universal of the OOP encapsulation and module standardization lifecycle or event mechanism becomes indispensable.
+
+## Propose the solution
```

Comment:
```The Solution``` (This entire article is your proposal, you don't have to re-state proposal here.)